### PR TITLE
[refactor] Vulkan program impl

### PR DESCRIPTION
**Pull request by AmesingFlank**
Add a `VulkanProgramImpl` as part of our efforts to cleanup `program.cpp` by making it less backend-dependent.

### CODE REVIEW

File: taichi/backends/vulkan/vulkan_program.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      std::vector<std::unique_ptr<SNodeTree>> &snode_trees_,
+      std::unordered_map<int, SNode *> &snodes,
+      SNodeGlobalVarExprMap &snode_to_glb_var_exprs_,
+      uint64 *result_buffer) override;
+
+  void synchronize() override {
+    vulkan_runtime_->synchronize();
+  }
+
+  std::unique_ptr<AotModuleBuilder> make_aot_module_builder() override {
```
Comment (ailzhang) [inline taichi/backends/vulkan/vulkan_program.h:42]: Btw there should be a corresponding change in program.cpp that redirects llvm/metal/vulkan calls to program_impl_. (I think I missed that part in my refactor sorry :P ). Would you mind helping changing it as well?

**Review by ailzhang — APPROVED** [review]
Awesome! Thanks!

File: taichi/backends/vulkan/vulkan_program.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  *result_buffer_ptr = (uint64 *)memory_pool->allocate(
+      sizeof(uint64) * taichi_result_buffer_entries, 8);
+  // doesn't do anything other than alloc result buffer. runtime is materialized
+  // together with snode tree.
+  // TODO: separate runtime materialization and tree materialization.
+}
+
+void VulkanProgramImpl::materialize_snode_tree(
+    SNodeTree *tree,
+    std::vector<std::unique_ptr<SNodeTree>> &snode_trees_,
```
[the remark this answers is not in the document]
Comment (AmesingFlank) [reply taichi/backends/vulkan/vulkan_program.cpp]: good idea.

File: taichi/backends/vulkan/vulkan_program.h (hunk 1 above)
Comment (AmesingFlank) [reply taichi/backends/vulkan/vulkan_program.h:41]: sure!

**Review by g1n0st — APPROVED** [review]

**Review by ailzhang — APPROVED** [review]
Thanks!

**Comment by k-ye** [comment]
Related issue: https://github.com/taichi-dev/taichi/issues/2736

**Pull request merged**